test/common/testlib.py

```diff
@@ -231,23 +231,30 @@ def focus(self, selector):
     def blur(self, selector):
         self.call_js_func('ph_blur', selector)
 
-    def key_press(self, keys):
-        for k in keys:
-            if k == "Backspace":
-                self.cdp.invoke("Input.dispatchKeyEvent", type="keyDown", windowsVirtualKeyCode=8)
-            elif k.isalnum():
-                self.cdp.invoke("Input.dispatchKeyEvent", type="char", text=k, key=k)
+    def key_press(self, keys, modifiers=0):
+        for key in keys:
+            args = { "type":"keyDown", "modifiers":modifiers };
+
+            # If modifiers are used we need to pass windowsVirtualKeyCode which is
+            # basically the asci decimal representation of the key
+            args["text"] = key
+            if (not key.isalnum()  and ord(key.upper()) < 32) or modifiers != 0:
```

Comment:
Why does it do the `upper()` here? If it's a control message, then upper() should either be a no-op, or damage the value. Also, please fix the duplicate space.

Reply:
Right, the upper is not needed here, my fault. 
But it's needed in next line, aka:
```args["windowsVirtualKeyCode"] = ord(key.upper())``` for the case where key is a letter and modifier are used. In the case that's not a letter. upper() will do nothing, it will not damage the value.
documentation [here](https://www.tutorialspoint.com/python/string_upper.htm) says 'This method returns a copy of the string in which all case-based characters have been uppercased.'